### ISSUE TITLE
:redir @+ does not work

### ISSUE THREAD

**Issue by ZyX-I** [neovim/neovim#1396]
Vanilla Vim used to contain

```C
	    if (ASCII_ISALPHA(*arg)
# ifdef FEAT_CLIPBOARD
		    || *arg == '*'
		    || *arg == '+'
# endif
		    || *arg == '"')
```

in `ex_redir()` in `ex_docmd.c`. The `FEAT_CLIPBOARD` part was removed and now `:redir @+` says that `@+` is invalid argument.

For this to work properly at least `write_reg_contents_ex` must support clipboard provider and it is currently missing `set_clipboard` call after `str_to_reg` call. It looks like just putting a single call where `may_set_selection` call originally was is sufficient.

**Comment by justinmk** [comment]
Seems like https://github.com/neovim/neovim/pull/1182 did not fix this issue. @bfredl

**Comment by bfredl** [comment]
There is so many places where register names are manually checked and special registers  treated specially in different ways, some of which must be kept for compability I guess (like `normal @@` and `let @@` are referring _different_ "default" registers) but here we could  just allow any writeable register I think.

**Comment by bfredl** [comment]
I will try to fix this in #2411

**Comment by bfredl** [comment]
Should be fixed now, please reopen if still problems
